karapace/compatibility/__init__.py

```diff
@@ -34,36 +37,62 @@ def is_transitive(self) -> bool:
         return self.value in TRANSITIVE_MODES
 
 
-class IncompatibleSchema(Exception):
-    pass
-
-
-def check_avro_compatibility(reader_schema, writer_schema) -> None:
+def check_avro_compatibility(reader_schema, writer_schema) -> SchemaCompatibilityResult:
     result = AvroChecker().get_compatibility(reader=reader_schema, writer=writer_schema)
     if (
         result.compatibility is SchemaCompatibilityType.incompatible
         and [SchemaIncompatibilityType.missing_enum_symbols] != result.incompatibilities
     ):
-        raise IncompatibleSchema(str(result.compatibility))
+        return result
+
+    return SchemaCompatibilityResult.compatible()
 
 
-def check_compatibility(source: TypedSchema, target: TypedSchema, compatibility_mode: CompatibilityModes) -> None:
+def check_jsonschema_compatibility(reader: Draft7Validator, writer: Draft7Validator) -> SchemaCompatibilityResult:
+    return jsonschema_compatibility(reader, writer)
+
+
+def check_compatibility(
+    source: TypedSchema, target: TypedSchema, compatibility_mode: CompatibilityModes
+) -> SchemaCompatibilityResult:
     if source.schema_type is not target.schema_type:
-        raise IncompatibleSchema(f"Comparing different schema types: {source.schema_type} with {target.schema_type}")
+        return SchemaCompatibilityResult.incompatible(
+            incompat_type=SchemaIncompatibilityType.type_mismatch,
+            message=f"Comparing different schema types: {source.schema_type} with {target.schema_type}",
+            location=[],
+        )
 
     if compatibility_mode is CompatibilityModes.NONE:
         LOG.info("Compatibility level set to NONE, no schema compatibility checks performed")
-        return
+        return SchemaCompatibilityResult.compatible()
 
     if source.schema_type is SchemaType.AVRO:
         if compatibility_mode in {CompatibilityModes.BACKWARD, CompatibilityModes.BACKWARD_TRANSITIVE}:
-            check_avro_compatibility(writer_schema=source.schema, reader_schema=target.schema)
+            result = check_avro_compatibility(reader_schema=target.schema, writer_schema=source.schema)
+
+        elif compatibility_mode in {CompatibilityModes.FORWARD, CompatibilityModes.FORWARD_TRANSITIVE}:
+            result = check_avro_compatibility(reader_schema=source.schema, writer_schema=target.schema)
+
+        elif compatibility_mode in {CompatibilityModes.FULL, CompatibilityModes.FULL_TRANSITIVE}:
+            result = check_avro_compatibility(reader_schema=target.schema, writer_schema=source.schema)
+            result = result.merged_with(check_avro_compatibility(reader_schema=source.schema, writer_schema=target.schema))
+
+    elif source.schema_type is SchemaType.JSONSCHEMA:
```

Comment:
Adding this is pretty much all that was needed to support jsonschema in the APIs, the rest of the code was already in place.